src/util/key.rs

```diff
@@ -212,32 +212,45 @@ impl Key for PublicKey {
         }
     }
 
+    /// If the key starts with `02`, `03` or `04` byte it makes the key ambigous, since it may
+    /// represent both a valid Secp and Taproot-encoded key. Thus, in case both representations
+    /// result in a valid keys the function will fail with [`std::io::ErrorKind::InvalidData`]
+    /// error wrapped into a [`std::io::Error`].
+    ///
+    /// There is a BIP-340 PR and discussion on how this can be prevented in a more standard way:
+    /// <https://github.com/bitcoin/bips/pull/1060>
     fn read_from<R: io::Read>(mut reader: R) -> Result<Self, io::Error> where Self: Sized {
-        // This is very imperfect algorithm that can end up in re-interpreting a valid Schnorr
-        // public key starting with `02`, `03` or `04` byte as another valid ECDSA key, which
-        // will lead to the shift in reading cursor...
-        // There is a BIP-340 PR and discussion on how this can be prevented:
-        // <https://github.com/bitcoin/bips/pull/1060>
 
         let mut bytes = [0; UNCOMPRESSED_PUBLIC_KEY_SIZE];
 
         reader.read_exact(&mut bytes[0..SCHNORRSIG_PUBLIC_KEY_SIZE])?;
+        let maybe_schnorr_key = schnorrsig::PublicKey::from_slice(&bytes[0..SCHNORRSIG_PUBLIC_KEY_SIZE]);
 
-        let (len, remaining_bytes) = if bytes[0] == 4 {
-            (UNCOMPRESSED_PUBLIC_KEY_SIZE, &mut bytes[SCHNORRSIG_PUBLIC_KEY_SIZE..UNCOMPRESSED_PUBLIC_KEY_SIZE])
-        } else if bytes[0] == 2 || bytes[0] == 3 {
-            (ECDSA_PUBLIC_KEY_SIZE, &mut bytes[SCHNORRSIG_PUBLIC_KEY_SIZE..ECDSA_PUBLIC_KEY_SIZE])
-        } else {
-            let key = schnorrsig::PublicKey::from_slice(&bytes[0..SCHNORRSIG_PUBLIC_KEY_SIZE])
-                .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
-            return Ok(PublicKey::Schnorr(key));
-        };
+        let len = { // Required by 1.29 borrow checker
+            let (len, remaining_bytes) = if bytes[0] == 4 {
+                (UNCOMPRESSED_PUBLIC_KEY_SIZE, &mut bytes[SCHNORRSIG_PUBLIC_KEY_SIZE..UNCOMPRESSED_PUBLIC_KEY_SIZE])
+            } else if bytes[0] == 2 || bytes[0] == 3 {
+                (ECDSA_PUBLIC_KEY_SIZE, &mut bytes[SCHNORRSIG_PUBLIC_KEY_SIZE..ECDSA_PUBLIC_KEY_SIZE])
```

Comment:
NACK

This code, as written, will always throw away the byte immediately following an x-only pubkey, and 50% of the time it will also incorrectly return an error.